core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakeAstBuilder.scala

```diff
@@ -56,31 +55,63 @@ class SnowflakeAstBuilder extends SnowflakeParserBaseVisitor[TreeNode] {
   }
 }
 
-class SnowflakeRelationBuilder extends SnowflakeParserBaseVisitor[Relation] {
-  override def visitObject_ref(ctx: SnowflakeParser.Object_refContext): Relation = {
+class SnowflakeRelationBuilder extends SnowflakeParserBaseVisitor[ir.Relation] {
+  override def visitObject_ref(ctx: SnowflakeParser.Object_refContext): ir.Relation = {
     val tableName = ctx.object_name().id_(0).getText
-    NamedTable(tableName, Map.empty, is_streaming = false)
+    ir.NamedTable(tableName, Map.empty, is_streaming = false)
   }
+
+  override def visitTable_source_item_joined(ctx: Table_source_item_joinedContext): ir.Relation = {
```

Comment:
maybe having a from_visitor will help navigate multiple type of expressions. 
It can be a subquery, join, cte etc. Thoughts?

Reply:
I was saying something similar 
https://github.com/apache/spark/blob/daf260f74e12fc5e9fad6091f6230e71a9e6c9c1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1046

Reply:
I haven't yet digested the whole setup yet. so I might be not thinking this through.

Reply:
Lemme give you some more context then. We are building `SnowflakeAstBuilder` as a subclass of `SnowflakeParserBaseVisitor` that ANTLR generates from the `SnowflakeParser.g4` grammar. It comes with a swath of `visit*` methods (such as the `visitTable_source_item_joined` above), basically one method for each rule in the grammar (here, the `table_source_item_joined` rule).  
So our job is to implement the `SnowflakeAstBuilder` visitor, whose structure is fully determined by the structure of the underlying grammar, in such a way that it outputs an AST that is similar to the one produced by catalyst's `AstBuilder` you mentioned above. 
So we have no choice but to cope with the fact that names and structures imposed by the Snowflake grammar (and reflected in the `SnowflakeApiBuilder`'s API) may sometimes be very different to the ones of our target AST, forcing us to produce code that isn't optimally readable/maintainable.
That being said, we should always keep that fact in mind and try to minimize the effects of this "structural discrepancy" as much as possible.
I hope this helps answering your question.

Reply:
Another thing to keep in mind is that for the moment, we focus on (quoting @nfx) "quantity over quality". Meaning that we accept to cut some corners in order to implement as much of the Snowflake grammar as quick as possible (although some "corners" remain "uncutable", such as maintaining good test coverage, etc).
So even though reviewers' comments about code quality are (as always) very much appreciated, please be aware that addressing those concerns might get temporarily postponed for the sake of "moving fast".

Reply:
@sundarshankar89 we need to strategically cut corners in the following two weeks and gather a backlog of what do we need to do to improve snowflake grammar. And then GDC can come in :)